test/Fencer/Rules/Test.hs

```diff
@@ -227,3 +229,66 @@ test_rulesLoadRulesReadPermissions =
   file2 = simpleRuleFile
     ("domain2" </> "config" </> "config.yml")
     domainDescriptorKeyText
+
+-- | test that faulty rules in reloading with 'loadRulesFromDirectory'
+-- are rejected and the valid ones that were previously loaded are
+-- kept instead.
+--
+-- This matches the behavior of @lyft/ratelimit@.
+test_rulesReloadRules :: TestTree
+test_rulesReloadRules =
+  testCase "Rules fail to reload for an invalid domain" $
+    Temp.withSystemTempDirectory "fencer-config" $ \tempDir -> do
+      rules <- writeAndLoadRules
+        (#ignoreDotFiles False)
+        (#root tempDir)
+        (#files files)
+      failures <- writeAndLoadRules
+        (#ignoreDotFiles False)
+        (#root tempDir)
+        (#files files')
+      case (rules, failures) of
+        (Left errs, _) ->
+          assertFailure
+            ("Expected domains, got failures: " ++
+             prettyPrintErrors (NE.toList errs))
+        (Right _, Right newDefinitions) ->
+          assertFailure
+            ("Expected failures, got domains: " ++ show newDefinitions)
+        (Right definitions, Left errs) -> do
+          let
+            sortedResult = sortOn domainDefinitionId <$> result
+            sortedDefs   = sortOn domainDefinitionId definitions
+          assertBool
+            ("Unexpected definitions! Expected: " ++ (show sortedResult) ++
+             "\nGot: " ++ show sortedDefs) $
+            ((==) `on` show) sortedResult (Right sortedDefs)
+
+          assertBool
+            ("Unexpected failure! Expected: " ++
+              (prettyPrintErrors . NE.toList $ expectedFailure) ++
+              "\nGot: " ++
+              (prettyPrintErrors . NE.toList $ errs)
+            ) $
+            ((==) `on` (fmap showError . NE.toList))
+              expectedFailure
+              (trimPath <$> errs)
+ where
+  files :: [RuleFile]
+  files =
+    [ simpleRuleFile ("domain1" </> "config.yml") domainDescriptorKeyValueText
+    , simpleRuleFile ("domain2" </> "config.yml") domainDescriptorKeyText ]
+  files' :: [RuleFile]
+  files' =
```

Comment:
Nit: I'd name it `filesFailure` or something.

(whenever I say "nit" I mean "feel free to do either way")

Reply:
Done!